pkg/sharding/listwatch.go

```diff
@@ -0,0 +1,88 @@
+package sharding
+
+import (
+	"crypto/md5"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+type shardedListWatch struct {
+	sharding *sharding
+	lw       cache.ListerWatcher
+}
+
+func NewShardedListWatch(shard, totalShards uint64, lw cache.ListerWatcher) cache.ListerWatcher {
+	// This is an "optimization" as this configuration means no sharding is to
+	// be performed.
+	if shard == 0 && totalShards == 1 {
+		return lw
+	}
+
+	return &shardedListWatch{sharding: &sharding{shard: shard, totalShards: totalShards}, lw: lw}
+}
+
+func (s *shardedListWatch) List(options metav1.ListOptions) (runtime.Object, error) {
+	list, err := s.lw.List(options)
+	if err != nil {
+		return nil, err
+	}
+	items, err := meta.ExtractList(list)
+	if err != nil {
+		return nil, err
+	}
+	res := &metav1.List{
+		Items: []runtime.RawExtension{},
+	}
+	for _, item := range items {
+		a, err := meta.Accessor(item)
+		if err != nil {
+			return nil, err
+		}
+		if s.sharding.keep(a) {
+			res.Items = append(res.Items, runtime.RawExtension{Object: item})
+		}
+	}
+
+	return res, nil
+}
+
+func (s *shardedListWatch) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	w, err := s.lw.Watch(options)
+	if err != nil {
+		return nil, err
+	}
+
+	return watch.Filter(w, watch.FilterFunc(func(in watch.Event) (out watch.Event, keep bool) {
+		a, err := meta.Accessor(in.Object)
+		if err != nil {
+			// TODO(brancz): needs logging
+			return in, true
+		}
+
+		return in, s.sharding.keep(a)
+	})), nil
+}
+
+type sharding struct {
+	shard, totalShards uint64
+}
+
+func (s *sharding) keep(o metav1.Object) bool {
+	return sum64(md5.Sum([]byte(o.GetUID())))%s.totalShards == s.shard
+}
+
+// sum64 sums the md5 hash to an uint64.
+func sum64(hash [md5.Size]byte) uint64 {
```

Comment:
If we use the golang/hash package, we could use the `Sum64` interface function instead, right?

```golang
type Hash64 interface {
        Hash
        Sum64() uint64
}
```

Reply:
Choosing the right hash function, woah ;-) I am literally not an expert, but I think the focus is here on good key distribution and speed. When we talk about "speed", the amount of data is not super critical here I guess as there are not that many objects to be calculated per second.

While looking at https://github.com/golang/groupcache/blob/c65c006176ff7ff98bb916961c7abbc6b0afc0aa/consistenthash/consistenthash.go#L42, crc32 might be fine, but https://medium.com/@dgryski/consistent-hashing-algorithmic-tradeoffs-ef6b8e2fcae8 points out some pros/cons.

Reply:
I chose this exact hashing mechanism as it's been well proven in the Prometheus `hashmod` system.